Fix brittle tests that depend on external sites

### Task Description

This is an extension of issue #1989 which fixed brittle pubmed tests, so that if pubmed happens to be down when tests are run, they will be gracefully skipped.
This is also related to #1501 where the first thing to try is to retry the download a couple of times. In particular, see this comment https://github.com/tripal/tripal/pull/1679#pullrequestreview-1718265923

The task is to both increase the robustness of downloads from external sites, and if the download still fails, process gracefully in automated tests

This is inspired today because the overnight PHPUnit tests are all failing.
Rerunning one of the failed tests completes without error, highly suggestive of this being a symptom of the external site being down.

### Branch Name

tv4g9-issue2143-fix-brittle-tests-depending-on-external-sites

### Example error messages from https://github.com/tripal/tripal/actions/runs/13612833349/job/38052135242
```
There were 3 failures:

1) Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
One publication should have been retrieved but was not
Failed asserting that 1 matches expected 0.

/var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:60

2) Drupal\Tests\tripal_chado\Functional\TaxonomyImporterTest::testTaxonomyImporterSimpleTest
No organism Arabidopsis thaliana found which should have been created
Failed asserting that '0' matches expected 1.

/var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TaxonomyImporterTest.php:61

3) Drupal\Tests\tripal_chado\Functional\TripalPubLibraryTest::testTripalPubLibraryTestSimpleTest
There should be more than 0 records found for this query
Failed asserting that null is greater than 0.

/var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TripalPubLibraryTest.php:65

--

2 tests triggered 6 PHP warnings:

1) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:287
fopen(): php_network_getaddresses: getaddrinfo for www.ncbi.nlm.nih.gov failed: Temporary failure in name resolution

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

* Drupal\Tests\tripal_chado\Functional\TripalPubLibraryTest::testTripalPubLibraryTestSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TripalPubLibraryTest.php:14

2) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:287
fopen(https://www.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=1&usehistory=y&term=%28%28Populus+%5BTitle%5D%29%29): Failed to open stream: php_network_getaddresses: getaddrinfo for www.ncbi.nlm.nih.gov failed: Temporary failure in name resolution

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

3) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:220
Trying to access array offset on int

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

* Drupal\Tests\tripal_chado\Functional\TripalPubLibraryTest::testTripalPubLibraryTestSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TripalPubLibraryTest.php:14

4) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:221
Trying to access array offset on int

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

* Drupal\Tests\tripal_chado\Functional\TripalPubLibraryTest::testTripalPubLibraryTestSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TripalPubLibraryTest.php:14

5) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:222
Trying to access array offset on int

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

* Drupal\Tests\tripal_chado\Functional\TripalPubLibraryTest::testTripalPubLibraryTestSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TripalPubLibraryTest.php:14

6) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:287
fopen(https://www.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=Pubmed&retmax=1&usehistory=y&term=%28%28Populus+trichocarpa+%5BTitle%2FAbstract%5D%29%29): Failed to open stream: php_network_getaddresses: getaddrinfo for www.ncbi.nlm.nih.gov failed: Temporary failure in name resolution

Triggered by:

* Drupal\Tests\tripal_chado\Functional\TripalPubLibraryTest::testTripalPubLibraryTestSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/TripalPubLibraryTest.php:14
```

Another one

1 test triggered 3 PHP warnings:

1) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:506
XMLReader::read(): /var/www/drupal/web/:6: parser error : Extra content at the end of the document

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

2) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:506
XMLReader::read(): </eFetchResult>

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

3) /var/www/drupal/web/modules/contrib/tripal/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php:506
XMLReader::read():                ^

Triggered by:

* Drupal\Tests\tripal_chado\Functional\PubSearchQueryImporterTest::testPubSearchQueryImporterSimpleTest
  /var/www/drupal/web/modules/contrib/tripal/tripal_chado/tests/src/Functional/Plugin/PubSearchQueryImporterTest.php:21

OK, but there were issues!
Tests: 587, Assertions: 25931, Warnings: 3, Incomplete: 10.
Error: Process completed with exit code 1.

--- a/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php
+++ b/tripal/src/Plugin/TripalPubLibrary/TripalPubLibraryPubMed.php
@@ -330,18 +330,12 @@ class TripalPubLibraryPubMed extends TripalPubLibraryBase {
     }
 
     usleep($sleep_time);  // 1/3 of a second delay, NCBI limits requests to 3 / second without API key
-    $rfh = fopen($query_url, "r");
-    if (!$rfh) {
+    $query_xml = $this->fileretriever->retrieveFileContents($query_url);
+    if (is_null($query_xml)) {
       $this->logger->error("Could not perform Pubmed query. Cannot connect to Entrez.");
       return FALSE;
     }
 
-    // retrieve the XML results
-    $query_xml = '';
-    while (!feof($rfh)) {
-      $query_xml .= fread($rfh, 255);
-    }
-    fclose($rfh);
     $xml = new \XMLReader();
     $xml->xml($query_xml);
 

--- a/tripal/tests/src/Kernel/Services/TripalFileRetriever/TripalFileRetrieverTest.php
+++ b/tripal/tests/src/Kernel/Services/TripalFileRetriever/TripalFileRetrieverTest.php
@@ -129,7 +129,8 @@ class TripalFileRetrieverTest extends TripalTestKernelBase {
       'https://github.com/vmasiufekxlkajfd.txt',
       [],
       [
-        'error_message' => 'Invalid file',
+        // Expect "Invalid file", but in rare cases when host is down can get "Invalid hostname"
+        'error_message' => 'Invalid',
         'has_content' => FALSE,
         'skip' => FALSE,
         'download_status' => FALSE,

